File: src/TypeFormatter/ObjectTypeFormatter.ts
import { Definition } from "../Schema/Definition.js";
import { SubTypeFormatter } from "../SubTypeFormatter.js";
import { AnyType } from "../Type/AnyType.js";
import { SymbolType } from "../Type/SymbolType.js";
import { BaseType } from "../Type/BaseType.js";
import { ObjectProperty, ObjectType } from "../Type/ObjectType.js";
import { UndefinedType } from "../Type/UndefinedType.js";
import { UnionType } from "../Type/UnionType.js";
import { TypeFormatter } from "../TypeFormatter.js";
import { getAllOfDefinitionReducer } from "../Utils/allOfDefinition.js";
import { derefType } from "../Utils/derefType.js";
import { preserveAnnotation } from "../Utils/preserveAnnotation.js";
import { removeUndefined } from "../Utils/removeUndefined.js";
import { StringMap } from "../Utils/StringMap.js";
import { uniqueArray } from "../Utils/uniqueArray.js";
import { NeverType } from "../Type/NeverType.js";

export class ObjectTypeFormatter implements SubTypeFormatter {
    public constructor(protected childTypeFormatter: TypeFormatter) {}

    public supportsType(type: BaseType): boolean {
        return type instanceof ObjectType;
    }

    public getDefinition(type: ObjectType): Definition {
        const types = type.getBaseTypes();
        if (types.length === 0) {
            return this.getObjectDefinition(type);
        }

        return types.reduce(getAllOfDefinitionReducer(this.childTypeFormatter), this.getObjectDefinition(type));
    }

    public getChildren(type: ObjectType): BaseType[] {
        const properties = type.getProperties();
        const additionalProperties: BaseType | boolean = type.getAdditionalProperties();

        const childrenOfBase = type
            .getBaseTypes()
            .reduce(
                (result: BaseType[], baseType) => [...result, ...this.childTypeFormatter.getChildren(baseType)],
                [],
            );

        const childrenOfAdditionalProps =
            additionalProperties instanceof BaseType ? this.childTypeFormatter.getChildren(additionalProperties) : [];

        const childrenOfProps = properties.reduce((result: BaseType[], property) => {
            const propertyType = property.getType();
            if (propertyType instanceof NeverType) {
                return result;
            }

            return [...result, ...this.childTypeFormatter.getChildren(propertyType)];
        }, []);

        const children = [...childrenOfBase, ...childrenOfAdditionalProps, ...childrenOfProps];

        return uniqueArray(children);
    }

    protected getObjectDefinition(type: ObjectType): Definition {
        let objectProperties = type.getProperties();
        const additionalProperties: BaseType | boolean = type.getAdditionalProperties();

        if (additionalProperties === false) {
            objectProperties = objectProperties.filter(
                (property) => !(derefType(property.getType()) instanceof NeverType),
            );
        }

        const preparedProperties = objectProperties.map((property) => this.prepareObjectProperty(property));

        const required = preparedProperties
            .filter((property) => property.isRequired())
            .map((property) => property.getName());

        const properties = preparedProperties.reduce((result: StringMap<Definition>, property) => {
            result[property.getName()] = this.childTypeFormatter.getDefinition(property.getType());
            return result;
        }, {});

        return {
            type: "object",
            ...(Object.keys(properties).length > 0 ? { properties } : {}),
            ...(required.length > 0 ? { required } : {}),
            ...(additionalProperties === true ||
            additionalProperties instanceof AnyType ||
            additionalProperties instanceof SymbolType
                ? {}
                : {
                      additionalProperties:
                          additionalProperties instanceof BaseType
                              ? this.childTypeFormatter.getDefinition(additionalProperties)
                              : additionalProperties,
                  }),
        };
    }

    protected prepareObjectProperty(property: ObjectProperty): ObjectProperty {
        const propertyType = property.getType();
        const propType = derefType(propertyType);

        if (propType instanceof UndefinedType) {
            return new ObjectProperty(property.getName(), propertyType, false);
        } else if (!(propType instanceof UnionType)) {
            return property;
        }

        const { newType: newPropType, numRemoved } = removeUndefined(propType);

        if (numRemoved == 0) {
            return property;
        }

        return new ObjectProperty(property.getName(), preserveAnnotation(propertyType, newPropType), false);
    }
}

File: internal/formatter/object_type_formatter.go
package formatter

import (
	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// ObjectTypeFormatter mirrors src/TypeFormatter/ObjectTypeFormatter.ts.
type ObjectTypeFormatter struct {
	childTypeFormatter TypeFormatter
}

func NewObjectTypeFormatter(childTypeFormatter TypeFormatter) *ObjectTypeFormatter {
	return &ObjectTypeFormatter{childTypeFormatter: childTypeFormatter}
}

func (f *ObjectTypeFormatter) SupportsType(t types.Type) bool {
	_, ok := t.(*types.ObjectType)
	return ok
}

func (f *ObjectTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	objectType := t.(*types.ObjectType)
	baseTypes := objectType.BaseTypes
	definition := f.getObjectDefinition(objectType)
	if len(baseTypes) == 0 {
		return definition
	}

	reducer := GetAllOfDefinitionReducer(f.childTypeFormatter)
	for _, baseType := range baseTypes {
		definition = reducer(definition, baseType)
	}
	return definition
}

func (f *ObjectTypeFormatter) GetChildren(t types.Type) []types.Type {
	objectType := t.(*types.ObjectType)
	properties := objectType.Properties
	additionalProperties := objectType.AdditionalProperties

	var children []types.Type

	for _, baseType := range objectType.BaseTypes {
		children = append(children, f.childTypeFormatter.GetChildren(baseType)...)
	}

	if additionalPropertiesType, ok := additionalProperties.(types.Type); ok {
		children = append(children, f.childTypeFormatter.GetChildren(additionalPropertiesType)...)
	}

	for _, property := range properties {
		propertyType := property.Type
		if types.IsNeverLike(propertyType) {
			continue
		}
		children = append(children, f.childTypeFormatter.GetChildren(propertyType)...)
	}

	return uniqueTypes(children)
}

func (f *ObjectTypeFormatter) getObjectDefinition(objectType *types.ObjectType) *schema.Definition {
	objectProperties := objectType.Properties
	additionalProperties := objectType.AdditionalProperties

	if additionalProperties == false {
		filtered := make([]*types.ObjectProperty, 0, len(objectProperties))
		for _, property := range objectProperties {
			if !types.IsNeverLike(types.DerefType(property.Type)) {
				filtered = append(filtered, property)
			}
		}
		objectProperties = filtered
	}

	preparedProperties := make([]*types.ObjectProperty, len(objectProperties))
	for i, property := range objectProperties {
		preparedProperties[i] = f.prepareObjectProperty(property)
	}

	var required []string
	for _, property := range preparedProperties {
		if property.Required {
			required = append(required, property.Name())
		}
	}

	properties := schema.NewProperties()
	for _, property := range preparedProperties {
		properties.Set(property.Name(), f.childTypeFormatter.GetDefinition(property.Type))
	}

	definition := &schema.Definition{Type: "object"}
	if properties.Len() > 0 {
		definition.Properties = properties
	}
	if len(required) > 0 {
		definition.Required = required
	}

	switch ap := additionalProperties.(type) {
	case bool:
		if !ap {
			definition.AdditionalProperties = false
		}
	case *types.AnyType, *types.SymbolType:
		// Omitted, like additionalProperties === true.
	case types.Type:
		definition.AdditionalProperties = f.childTypeFormatter.GetDefinition(ap)
	}

	return definition
}

func (f *ObjectTypeFormatter) prepareObjectProperty(property *types.ObjectProperty) *types.ObjectProperty {
	propertyType := property.Type
	propType := types.DerefType(propertyType)

	if _, ok := propType.(*types.UndefinedType); ok {
		return types.NewObjectProperty(property.Name(), propertyType, false)
	}
	unionType, ok := propType.(*types.UnionType)
	if !ok {
		return property
	}

	numRemoved, newPropType := types.RemoveUndefined(unionType)
	if numRemoved == 0 {
		return property
	}

	return types.NewObjectProperty(property.Name(), types.PreserveAnnotation(propertyType, newPropType), false)
}
